scipy/special/_lambertw.py

```diff
@@ -141,4 +143,7 @@ def lambertw(z, k=0, tol=1e-8):
     >>> -lambertw(-np.log(0.5)) / np.log(0.5)
     (0.64118574450498589+0j)
     """
+    # TODO: special expert should inspect this
+    # interception; better place to do it?
+    k = np.asarray(k, dtype=np.dtype("long"))
```

Comment:
According to the function signature in `special/functions.json` and [`lambertw.h`](https://github.com/scipy/scipy/blob/main/scipy/special/special/lambertw.h#L62) `k` is definitely expected to be a `long`. Therefore this change seems to be correct to me. It's easier to do this kind of manipulation in Python rather than in lower level code.

**not a special expert.

Reply:
That it's only an issue for `lambertw` and not any of the other special ufuncs which have an input of type long seems to relate to `lambertw` being the only of these written in C++, the others are all written in Cython.

I munged `functions.json` to get the list. There's 32 such functions.

```
[['_lambertw', '_special.h++', 'Dld->D'],
 ['_spherical_in', '_spherical_bessel.pxd', 'lD->D'],
 ['_spherical_in', '_spherical_bessel.pxd', 'ld->d'],
 ['_spherical_in_d', '_spherical_bessel.pxd', 'lD->D'],
 ['_spherical_in_d', '_spherical_bessel.pxd', 'ld->d'],
 ['_spherical_jn', '_spherical_bessel.pxd', 'lD->D'],
 ['_spherical_jn', '_spherical_bessel.pxd', 'ld->d'],
 ['_spherical_jn_d', '_spherical_bessel.pxd', 'lD->D'],
 ['_spherical_jn_d', '_spherical_bessel.pxd', 'ld->d'],
 ['_spherical_kn', '_spherical_bessel.pxd', 'lD->D'],
 ['_spherical_kn', '_spherical_bessel.pxd', 'ld->d'],
 ['_spherical_kn_d', '_spherical_bessel.pxd', 'lD->D'],
 ['_spherical_kn_d', '_spherical_bessel.pxd', 'ld->d'],
 ['_spherical_yn', '_spherical_bessel.pxd', 'lD->D'],
 ['_spherical_yn', '_spherical_bessel.pxd', 'ld->d'],
 ['_spherical_yn_d', '_spherical_bessel.pxd', 'lD->D'],
 ['_spherical_yn_d', '_spherical_bessel.pxd', 'ld->d'],
 ['eval_chebyc', 'orthogonal_eval.pxd', 'ld->d'],
 ['eval_chebys', 'orthogonal_eval.pxd', 'ld->d'],
 ['eval_chebyt', 'orthogonal_eval.pxd', 'ld->d'],
 ['eval_chebyu', 'orthogonal_eval.pxd', 'ld->d'],
 ['eval_gegenbauer', 'orthogonal_eval.pxd', 'ldd->d'],
 ['eval_genlaguerre', 'orthogonal_eval.pxd', 'ldd->d'],
 ['eval_hermite', 'orthogonal_eval.pxd', 'ld->d'],
 ['eval_hermitenorm', 'orthogonal_eval.pxd', 'ld->d'],
 ['eval_jacobi', 'orthogonal_eval.pxd', 'lddd->d'],
 ['eval_laguerre', 'orthogonal_eval.pxd', 'ld->d'],
 ['eval_legendre', 'orthogonal_eval.pxd', 'ld->d'],
 ['eval_sh_chebyt', 'orthogonal_eval.pxd', 'ld->d'],
 ['eval_sh_chebyu', 'orthogonal_eval.pxd', 'ld->d'],
 ['eval_sh_jacobi', 'orthogonal_eval.pxd', 'lddd->d'],
 ['eval_sh_legendre', 'orthogonal_eval.pxd', 'ld->d']]
```

Is Cython doing something to handle this? The plan though is to have all of these translated into C++ before we branch for 1.13. `lambertw` and the spherical bessel functions have convenient wrappers so they can support kwargs with default arguments, but the orthogonal polynomial evaluation functions are called directly. We could write wrappers for those and do the conversion in Python like you've done here, but the functions would lose `ufunc` methods and attributes, which some might complain about.

 I think we could change the [ufunc generation infrastructure](
https://github.com/scipy/scipy/blob/e72de0714f0ad6e9b93c9107be84ee68ef70a5db/scipy/special/_generate_pyx.py#L244-L277)  to explicitly use `int64` types instead of `long` but I don't know offhand how to put everything together to make that work. I think it's fine to just do the conversion in Python for now, and I'll work on updating the ufunc infrastructure before rewriting `orthogonal_eval.pxd`.